### ISSUE TITLE
[C#] Non-windows performance

### ISSUE THREAD

**Issue by JeremyTCD** [microsoft/FASTER#429]
Hey! This issue continues the discussion [here](https://github.com/JeringTech/KeyValueStore/issues/1).  

I'm having trouble with read and write performance on non-windows platforms. [Repro](https://github.com/JeremyTCD/Repro.Faster.NonWindowsPerformance). Usage:

- Clone
- `cd Repro.Faster.NonWindowsPerformance`
- Open a WSL terminal
- `dotnet run --configuration "Release"`

On my machine, inserting and reading 508 records:
```
Profiling MinimalKVStore:
    Inserting 508 records ...
    Insertion complete in 22 ms
    Reading 508 records ...
    Reads complete in 13 ms
    Verifying read results ...
    Results verified
```

Inserting and reading  509 records:

```
Profiling MinimalKVStore:
    Inserting 509 records ...
    Insertion complete in 22 ms
    Reading 509 records ...
```

At 509 records, all CPU's run at 100% utilization once reads begin. In the run that produced the snippet above, reads hadn't completed after 20 minutes so I killed the process.

The above results are deterministic (on my machine) - 508 records always completes quickly while 509 records always hangs. Perhaps 508 records fit in memory.  

Increasing the number of records to 1000 significantly slows down writes as well (~ 8 s).

The same profiling operation is stable and performant on Windows.  

For quick reference, the store:

```csharp
public class ObjLogMinimalKVStore : IMinimalKVStore
{
    private readonly SimpleFunctions<int, int, Empty> _simpleFunctions = new();
    private readonly ConcurrentQueue<ClientSession<int, int, int, int, Empty, SimpleFunctions<int, int, Empty>>> _sessionPool = new();
    private readonly FasterKV<int, int> _kvStore;
    private readonly FasterKV<int, int>.ClientSessionBuilder<int, int, Empty> _clientSessionBuilder;

    public int NumSessions => _sessionPool.Count;

    public ObjLogMinimalKVStore()
    {
        // Settings
        string logDirectory = Path.Combine(Path.GetTempPath(), "FasterLogs");
        string logFileName = Guid.NewGuid().ToString();
        var logSettings = new LogSettings
        {
            LogDevice = Devices.CreateLogDevice(Path.Combine(logDirectory, $"{logFileName}.log"), deleteOnClose: true),
            ObjectLogDevice = Devices.CreateLogDevice(Path.Combine(logDirectory, $"{logFileName}.obj.log"), deleteOnClose: true),
            PageSizeBits = 12,
            MemorySizeBits = 13
        };

        // Create store
        _kvStore = new(1L << 20, logSettings);
        _clientSessionBuilder = _kvStore.For(_simpleFunctions);
    }

    public void Upsert(int key, int value)
    {
        var session = GetPooledSession();
        session.Upsert(key, value);
        _sessionPool.Enqueue(session);
    }

    public async Task<(Status, int?)> ReadAsync(int key)
    {
        var session = GetPooledSession();
        (Status, int) result = (await session.ReadAsync(key).ConfigureAwait(false)).Complete();
        _sessionPool.Enqueue(session);

        return result;
    }

    private ClientSession<int, int, int, int, Empty, SimpleFunctions<int, int, Empty>> GetPooledSession()
    {
        if (_sessionPool.TryDequeue(out ClientSession<int, int, int, int, Empty, SimpleFunctions<int, int, Empty>>? result))
        {
            return result;
        }

        return _clientSessionBuilder.NewSession<SimpleFunctions<int, int, Empty>>();
    }

    public void Dispose()
    {
        foreach (var session in _sessionPool)
        {
            session.Dispose();
        }

        _kvStore.Dispose();
    }
}
```

The profiling logic:

```csharp
private static async Task ProfileStore(IMinimalKVStore minimalKVStore)
{
    Console.WriteLine($"Profiling {minimalKVStore.GetType().Name}:");
    Stopwatch stopWatch = new Stopwatch();
    const int numOperations = 509;

    // Insert
    Console.WriteLine($"    Inserting {numOperations} records ...");
    stopWatch.Start();
    Parallel.For(0, numOperations, key => minimalKVStore.Upsert(key, key));
    stopWatch.Stop();
    Console.WriteLine($"    Insertion complete in {stopWatch.ElapsedMilliseconds} ms");

    // Read
    ConcurrentDictionary<long, Task<(Status, int?)>> readTasks = new();
    Console.WriteLine($"    Reading {numOperations} records ...");
    stopWatch.Restart();
    Parallel.For(0, numOperations, key => readTasks.TryAdd(key, minimalKVStore.ReadAsync(key)));
    await Task.WhenAll(readTasks.Values).ConfigureAwait(false);
    stopWatch.Stop();
    Console.WriteLine($"    Reads complete in {stopWatch.ElapsedMilliseconds} ms");

    // Verify
    Console.WriteLine("    Verifying read results ...");
    Parallel.For(0, numOperations, key =>
    {
        (Status status, int? result) = readTasks[key].Result;
        Assert.Equal(Status.OK, status);
        Assert.Equal(key, result);
    });

    Console.WriteLine("    Results verified");

    minimalKVStore.Dispose();
}
```

Context:

- Faster 1.8.4
- Ubuntu 18.04

**Comment by badrishc** [comment]
Some progress: its not Win vs Linux. Its the default use of ManagedLocalStorageDevice instead of LocalStorageDevice. It repros on Windows by replacing CreateDevice with: new ManagedLocalStorageDevice().

For Upsert (1000 records), it seems to be that the continuation of the WriteAsync in MLSD rarely gets scheduled because all threads are looping on BlockAllocate trying to add new data to the tail. There is a Thread.Yield in BlockAllocate, but that seems insufficient to prioritize the IO callback.

**Comment by badrishc** [comment]
For Upserts, do not use `Parallel.For`. You are flooding FASTER with many threads spinning on allocation on the tail, causing thread pool starvation. As a result, the page flush continuations do not get scheduled. Instead, just do:

```cs
            for (int i=0; i< numOperations; i++)
                minimalKVStore.Upsert(i, i);
```

**Comment by badrishc** [comment]
Similarly for reads, this will fix the problem:

```cs
            for (int i = 0; i < numOperations; i++)
                readTasks.TryAdd(i, minimalKVStore.ReadAsync(i));
```

`Parallel.For` is to be avoided as it saturates the thread pool.

**Comment by JeremyTCD** [comment]
Hmm what if I'm building an application that is multi-threaded by design. E.g. an Asp.Net application handling concurrent requests on separate threads.

**Comment by badrishc** [comment]
We just do `stream.WriteAsync(...).ContinueWith(...)` and the ContinueWith is not getting scheduled because of so many threads actively working.

**Comment by badrishc** [comment]
For reads, you can continue to use parallel.for and fix the issue by avoiding disk throttling where you create the device:

```cs
            ThreadPool.GetMaxThreads(out int threads, out _);
            logSettings.LogDevice.ThrottleLimit = threads;
            logSettings.ObjectLogDevice.ThrottleLimit = threads;
```

**Comment by JeremyTCD** [comment]
Thanks for investigating. I couldn't get this to work:
```csharp
ThreadPool.GetMaxThreads(out int threads, out _);
logSettings.LogDevice.ThrottleLimit = threads;
logSettings.ObjectLogDevice.ThrottleLimit = threads;
```

This worked though:
```csharp
int processorCount = Environment.ProcessorCount;
...
Parallel.For(0, numOperations, new ParallelOptions() { MaxDegreeOfParallelism = processorCount }, key => minimalKVStore.Upsert(key, key));
...
Parallel.For(0, numOperations, new ParallelOptions() { MaxDegreeOfParallelism = processorCount }, key => readTasks.TryAdd(key, minimalKVStore.ReadAsync(key)));
```
Output:
```
Profiling ObjLogMinimalKVStore:
    Processor count: 8
    Inserting 10000 records ...
    Insertion complete in 129 ms
    Reading 10000 records ...
    Reads complete in 155 ms
    Verifying read results ...
    Results verified
```

Increasing `processorCount` by 1 causes reads and writes to "hang". To be honest this seems pretty dangerous. Accidentally performing 9 concurrent reads will cause a whole server to seize up.  

I'm using `Parallel.For` to simulate parallelism in applications. Will have to do a bit of research to figure out how the limit on parallelization can be enforced in general applications. `ThreadPool.SetMaxThreads` doesn't seem to do the job. Thoughts?

**Comment by badrishc** [comment]
You should set the throttle limit to max threads immediately after creating the devices. That worked for me for reads. Try setting the limit to very large like int.MaxValue

**Comment by badrishc** [comment]
> Can't get it to work on my machine. Will look into it again later on.

Try on Windows with `new ManagedLocalStorageDevice` instead of Linux and let me know if that works.

**Comment by badrishc** [comment]
Also, when the read hangs or slows down, see where most of the threads are (by pausing debugger) and seeing the threads in VS. You can see the call stack

**Comment by JeremyTCD** [comment]
Okay I've tried `new ManagedLocalStorageDevice` on windows, hangs here:

![deadlock](https://user-images.githubusercontent.com/11733898/110943127-cc7ba680-8375-11eb-808a-a0880f4d7779.png)

https://github.com/microsoft/FASTER/blob/c2ba0c4b73d1c2b8705d9dcd1b8c317c305664a1/cs/src/core/Index/FASTER/FASTERImpl.cs#L1736-L1750

**Comment by JeremyTCD** [comment]
`TryAllocate` is returning 0 at lines 780 and 796:

https://github.com/microsoft/FASTER/blob/c2ba0c4b73d1c2b8705d9dcd1b8c317c305664a1/cs/src/core/Allocator/AllocatorBase.cs#L775-L798

**Comment by badrishc** [comment]
What you showed would be during upsert, that's expected. Above, I am asking about Reads. There should be no allocate during reads. Trying to see what the fix I provided above is not helping for reads.

**Comment by JeremyTCD** [comment]
It's not getting to reads since upserts block indefinitely.

**Comment by badrishc** [comment]
I know, just fix the upsert as I said above (removing Parallel.For  only there), then try. We are diagnosing reads here, not upserts. The point I am making is: Reads should be fixed wrt Parallel.For by the throttle fix!

**Comment by badrishc** [comment]
See thread above, I mentioned "for reads" in my throttle suggestion messages above. Upserts are orthogonal, we worry about them later.

**Comment by JeremyTCD** [comment]
Ah my bad. Yeah reads work on windows now. Reads still hang on Linux.  

To be honest, I don't really have the bandwidth to help figure out these issues at the moment - after freezing the Faster KeyValueStore wrapper project, I put together a non-Faster alternative, so this is not a priority. 

I opened this issue only because of the discussion in the earlier thread. No pressure on my end to fix these issues: if the scenario `Parallel.For` simulates isn't what this library is meant for, that's fine, I can tweak the tests and benchmarks in the KeyValueStore project lmk.  

I've updated the KeyValueStore project's readme. Removed my misassumption about the cause of slow operations on Linux, linked to this issue instead.
